Add option [-V|--version] to emit version string

TerraTech opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->
Adds new option ```-V``` and ```--version``` to emit a version string
```sh
$ zfs --version
zfs-0.8.0-rc3-136-g470ba628f-dirty
zfs-kmod-0.7.13-r0-gentoo
```

NOTE: This collides with '-V' being used in: https://github.com/zfsonlinux/zfs/pull/8418 and I've made comment about that within the referred PR.

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
It is nice to be able to query a program's version.
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->
This is currently in RFC stasis for:
1. option flag selection
1. version string formatting
1. centralize and apply to all other binaries

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
Through usage.
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

ahrens commented:
It looks like this prints the version of the userland /sbin/zfs.  Since we support a mismatched kernel module and userland bits, I think that ideally this would print both the userland and kernel versions.  It would also be nice to print (perhaps in an extra-verbose mode) the ZFS_META_GITREV.

ahrens commented:
FYI, once we settle on what we want to do for ZoL, let's also bring this up on developer@open-zfs.org, so that folks from other platforms are aware of this proposed new feature.  They may have input on how the version should be expressed on different platforms (or perhaps ignored, since typically ZFS is part of the overall OS in FreeBSD and illumos, not packaged/delivered separately).

kpande commented:
> ZFS is part of the overall OS in FreeBSD

[false](https://www.freshports.org/sysutils/zol-kmod/).

TerraTech commented:
Overall, I'm just approaching this from the angle of gettings its foot in the door and at least making it available.  I believe this is to be a very benign feature addition.  Once it goes mainline, then I'm sure it can be tweaked (or bikeshedded) from there.  Also, is there a current method to query the running kernel module's version?

ahrens commented:
@TerraTech I am also a fan of getting the minimum useful thing done.  I would be OK with leaving ZFS_META_GITREV to a future extension (and I could imagine others too, like pretty-printing `/sys/module/zfs/features.*`).  But I feel it would be misleading to consider only the userland software version, since that may not represent the capabilities of the system (which are more closely defined by the kernel module).

TerraTech commented:
@ahrens is there a portable way to retrieve the running kernel module?  I see ```/sys/module/zfs/version``` but I don't think that would be portable in *BSD land.

Also, I just looked on my system (Gentoo-4.19.27, zfs-kmod-0.7.13) and I didn't see ```/sys/module/zfs/features.*```

ahrens commented:
> is there a portable way to retrieve the running kernel module

Not that I know of.  We could add an IOCTL to get the kernel version, but that probably isn't strictly necessary, since as you noted we can use `/sys/module/zfs/version`.  ZFS_META_* is also ZoL-specific.

> Also, I just looked on my system (Gentoo-4.19.27, zfs-kmod-0.7.13) and I didn't see /sys/module/zfs/features.*

It's new in 0.8.

TerraTech commented:
I have force pushed a reworked version and added 3 new functions to libzfs_util.c for version handling.

If it looks ok, then I will work on wiring it into the other commands.

TerraTech commented:
I've fixed up the buildbot errors, and have reformatted how the version string is emitted which should be cleaner now.

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,91 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+	(void) snprintf(version, 256, "%s%s",
+	    ZFS_META_GITREV, ZFS_VERSION_EXTRA);
+
+	return version;
```
[cstyle nit] add parentheses around return value (a few more places too)

TerraTech replied:
Will fix them up.

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,91 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+{
+	char *f_zversion = "/sys/module/zfs/version";
+	char *notfound = "zfs-kmod version not found";
+	char *version = safe_malloc(64);
```
Why is this buffer size different than in zfs_version_userland()?

TerraTech replied:
I generally try to be frugal with memory allocations and I noticed the string from "/sys/module/zfs/version" wasn't very long (17) on my system.  What would you recommend?

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,91 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
 }
+
+static void
+nomem(void)
```
I think that this code belongs in `zfs_main.c`, rather than libzfs.  If we keep it in libzfs, we should probably change it to not `exit()` or `printf()`, instead returning error codes and strings.

TerraTech replied:
I was a little surprised this wasn't centralized as I saw 'nomem()' and 'safe_malloc()' in various C files.  I think this is something that needs some tending to and try to DRY it everywhere.  I will leave this alone for now until consensus.

ahrens replied:
It is centralized, at least to a degree.  See zfs_alloc() and no_memory() in this file.

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,91 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+
+#ifndef ZFS_VERSION_EXTRA
+#define ZFS_VERSION_EXTRA ""
+#endif
```
Do we really need to introduce another type of version name, in addition to ZFS_META_VERSION, ZFS_META_RELEASE, ZFS_META_NAME, and ZFS_META_ALIAS?  Can we use one of those for this purpose instead?

TerraTech replied:
I didn't want to hijack an existing define, so I added one for those that would look at the code and use it if they wanted to.

behlendorf replied:
The `ZFS_META_RELEASE` field is intended for use by downstream packages, so I don't think you'd be hijacking an existing define.  You should need only to update the META file when building.

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,91 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+#define ZFS_VERSION_EXTRA ""
+#endif
+/*
+ * Returns the zfs userland version with ZFS_META_RELEASE and ZFS_META_GITREV
```
This comment doesn't seem to match the code, which returns ZFS_META_GITREV and ZFS_VERSION_EXTRA.  But I think we should be printing something like ZFS_META_RELEASE (perhaps in addition to ZFS_META_GITREV).

TerraTech replied:
From the looks of it, ZFS_META_GITREV pretty much was an all inclusive version string and I had removed the others and just used it exclusively.  Did I miss something?

ahrens replied:
I think you're right, ZFS_META_GITREV incorporates the others, so I think you're good, just update the comment.

TerraTech commented:
@ahrens Thank you for your review and input, much appreciated.

I've force-pushed the next iteration and tried to streamline it a bit by making it more conducive to being in a library with regard to error handling.  I also dropped ```nomem()``` function and cleaned it up as well.

behlendorf commented on cmd/zfs/zfs_main.c:
```diff
@@ -8114,6 +8114,15 @@ main(int argc, char **argv)
+	    (strcmp(cmdname, "--version") == 0)) {
+		zfs_version_print();
+		return (0);
+	}
```
Can you add this same special case to `cmd/zpool/zpool_main.c`.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,99 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+zfs_version_print(void)
+{
+	char *zver_userland;
+	char *zver_kernel;
```
Since this is all in user space, and the version strings are relatively short I don't think it would be unreasonable to put them on the stack.  This would ensure the allocation couldn't fail and remove the need for the `zfs_malloc()` function and a lot of the error handling.

```
int zfs_version_userland(char *version, int len);
int zfs_version_userland(char *version, int len);
```

TerraTech replied:
I'm just going to drop the concept of ZFS_VERSION_EXTRA altogether.  I wanted to suffix the version string, but trying to use ZFS_META_RELEASE will wedge in whatever I add into the middle of ZFS_META_GITREV.  At this point, it's just not worth it.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,99 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+}
+
+/*
+ * Returns the zfs kernel version read from "/sys/module/zfs/version"
```
You can use `ZFS_SYSFS_DIR` for the `/sys/module/zfs/` component.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1941,99 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
 }
+
+/*
+ * Utility function to guarantee malloc() success.
```
This comment no longer appears to be true.

TerraTech commented:
@behlendorf I've force pushed another iteration and have tried to simplify it taking into account your feedback.

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1942,61 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+{
+	(void) strlcpy(version, ZFS_META_GITREV, len);
+
+	return (1);
```
might as well make it return (void) since it only ever returns 1, and the caller ignores it.

TerraTech replied:
For whatever reason, I was just making both function prototypes the same.  I will change it to (void).

ahrens commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1942,61 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+/*
+ * Prints both zfs userland and kernel versions
+ */
+int
```
looks like callers don't check the return value

TerraTech commented:
@ahrens force-pushed, hopefully final version.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -9252,6 +9252,15 @@ main(int argc, char **argv)
+	    (strcmp(cmdname, "--version") == 0)) {
+		zfs_version_print();
+		return (0);
+	}
```
Let's add the new option to the `zfs.8` and `zpool.8` man pages.  Something like this:

```diff
diff --git a/man/man8/zpool.8 b/man/man8/zpool.8
index 7ba8a5b..fcef3c2 100644
--- a/man/man8/zpool.8
+++ b/man/man8/zpool.8
@@ -35,7 +35,7 @@
 .Nd configure ZFS storage pools
 .Sh SYNOPSIS
 .Nm
-.Fl ?
+.Fl ?V
 .Nm
 .Cm add
 .Op Fl fgLnP
@@ -909,6 +909,11 @@ The following subcommands are supported:
 Displays a help message.
 .It Xo
 .Nm
+.Fl V -version
+.Xc
+Displays the ZFS version information.
+.It Xo
+.Nm
 .Cm add
 .Op Fl fgLnP
 .Oo Fl o Ar property Ns = Ns Ar value Oc
```

TerraTech replied:
Thank you for that, I've added to both zfs.8 and zpool.8

TerraTech commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1942,59 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+
+/*
+ * Fill given version buffer with zfs kernel version read from ZFS_SYSFS_DIR
+ * Return 0 if error with errno set
```
Which would you prefer?
1. Returns 0 on success, and -1 on error (with errno set).
1. Returns 0 on success, and errno on error.

I see # 1 style throughout the code.

behlendorf replied:
Let's go with # 1 since this is what's done for most of the user space code.

TerraTech commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1942,59 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+	if ((fd = open(ZFS_SYSFS_DIR "/version", O_RDONLY)) == -1)
+		return (0);
+
+	if ((rlen = read(fd, version, len)) == -1)
```
ah, right - good catch.

TerraTech commented:
force-pushed (again).
@ahrens @behlendorf I'm really sorry.  I didn't mean for this to consume/distract a lot of your time for what seemed to be very low hanging fruit and easy to implement.

ahrens commented on man/man8/zpool.8:
```diff
@@ -909,6 +909,11 @@ The following subcommands are supported:
 .Nm
+.Fl V -version
+.Xc
+Displays the ZFS version information
```
Let's describe this a bit more precisely, since there's potential confusion with the versioning described by "zpool upgrade".  Specifically, this is telling about the software version, not to be confused with the on-disk format.

TerraTech replied:
Perhaps:
_Displays the ZFS program and kernel module version information_
?

ahrens replied:
"ZFS program" could refer to the `zfs program` subcommand.  How about `Displays the software version of the zfs userland utility and the zfs kernel module.`  Or maybe instead of `software version`, we could say `software release`?

behlendorf replied:
My feeling is `"Displays the software version of the zfs userland utility and the zfs kernel module"` is the better wording and is consistent with the option name.  Though, I'd OK with `"software release"` as well.

TerraTech replied:
done

ahrens commented on man/man8/zfs.8:
```diff
@@ -2453,6 +2453,11 @@ original form.
 .Nm
+.Fl V -version
+.Xc
+Displays the ZFS version information.
```
Why are we doing this as a flag, rather than a new subcommand?  You can't do `zfs --version create ...` or any other subcommands, can you?

TerraTech replied:
Because most that use GNU (and other) software use the de facto standard method of:
$ program --version
(that is always the first thing I usually try)

just like:
$ program --help

ahrens replied:
That makes sense.  I think we should consider making a `zfs version` subcommand, and have the `--version`/`-V` be a shortcut for running `zfs version`.

TerraTech replied:
done

ahrens commented on cmd/zfs/zfs_main.c:
```diff
@@ -8114,6 +8114,15 @@ main(int argc, char **argv)
+	 * Special case '-V|--version'
+	 */
+	if ((strcmp(cmdname, "-V") == 0) ||
+	    (strcmp(cmdname, "--version") == 0)) {
```
Seems like this should appear in the usage message as well.  How to do that depends on if we convert this to a subcommand or not.

TerraTech replied:
I would defer to yourself and @behlendorf on making that determination.

rlaager replied:
The usage doesn't show `--help`, so I don't think it needs `-V` or `--version` either. It seems like the "main" way is the `version` subcommand, which is listed in the usage. So I think this is fine as is, but I'm not going to go so far as to mark this resolved.

TerraTech commented:
force-pushed updated version
1. add zpool version
1. add zfs version
1. better return handling in libzfs_util.c
1. updated zfs.8 and zpool.8
NOTE: I had a heck of a time bending groff_mdoc to my will in how I wanted to format the 'zfs -V --version \n zfs version \n ....' lines.  A fair amount of trial and error before it finally settled in and behaved.  If there is a better way, please let me know.  It was my first real exposure to mdoc and I spent a lot of time trying to absorb groff_mdoc.7

behlendorf commented on cmd/zfs/zfs_main.c:
```diff
@@ -8059,6 +8065,18 @@ zfs_do_project(int argc, char **argv)
+zfs_do_version(int argc, char **argv)
+{
+	if (zfs_version_print() == -1)
+		exit(errno);
```
`zfs(8)` should exit with `1` on error, let's `return (1);`.

```
EXIT STATUS
     The zfs utility exits 0 on success, 1 if an error occurs, and 2 if invalid
     command line options were specified.
```

TerraTech replied:
will do

TerraTech replied:
done

behlendorf commented on cmd/zfs/zfs_main.c:
```diff
@@ -8114,6 +8132,12 @@ main(int argc, char **argv)
+	 * Special case '-V|--version'
+	 */
+	if ((strcmp(cmdname, "-V") == 0) || (strcmp(cmdname, "--version") == 0))
+		(void) zfs_do_version(argc, argv);
```
`return (zfs_do_version(argc, argv));` with the changes above.

TerraTech replied:
done

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -9222,6 +9229,18 @@ find_command_idx(char *command, int *idx)
+	if (zfs_version_print() == -1)
+		exit(errno);
+
+	exit(0);
```
See `zfs_do_version()` comments.

TerraTech replied:
done

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -9252,6 +9271,12 @@ main(int argc, char **argv)
+	 * Special case '-V|--version'
+	 */
+	if ((strcmp(cmdname, "-V") == 0) || (strcmp(cmdname, "--version") == 0))
+		(void) zpool_do_version(argc, argv);
```
`See zfs_do_version() comments.`

TerraTech replied:
done

behlendorf commented on man/man8/zfs.8:
```diff
@@ -38,7 +38,9 @@
-.Fl ?
+.Fl ?V
+.Nm
+.Cm version
```
nit: putting this at the beginning made sense when there wasn't a `zfs version` subcommand.  But since you've added one it should be placed in alphabetical order, which will be at the very end it seems.

behlendorf commented on man/man8/zpool.8:
```diff
@@ -35,7 +35,9 @@
-.Fl ?
+.Fl ?V
+.Nm
+.Cm version
```
I always end up fighting with mandoc too, but I believe this is what we want.  This puts the `zpool -V --version` alias first, and the `zpool version` subcommand with the others in alphabetical order.

```diff
diff --git a/man/man8/zpool.8 b/man/man8/zpool.8
index 7ba8a5bae9..410f34afee 100644
--- a/man/man8/zpool.8
+++ b/man/man8/zpool.8
@@ -35,7 +35,7 @@
 .Nd configure ZFS storage pools
 .Sh SYNOPSIS
 .Nm
-.Fl ?
+.Fl ?V
 .Nm
 .Cm add
 .Op Fl fgLnP
@@ -210,6 +210,8 @@
 .Cm upgrade
 .Op Fl V Ar version
 .Fl a Ns | Ns Ar pool Ns ...
+.Nm
+.Cm version
 .Sh DESCRIPTION
 The
 .Nm
@@ -909,6 +911,13 @@ The following subcommands are supported:
 Displays a help message.
 .It Xo
 .Nm
+.Fl V, -version
+.Xc
+An alias for the
+.Nm zpool Cm version
+subcommand.
+.It Xo
+.Nm
 .Cm add
 .Op Fl fgLnP
 .Oo Fl o Ar property Ns = Ns Ar value Oc
@@ -2441,6 +2450,14 @@ flag is specified, no features will be enabled on the pool.
 This option can only be used to increase the version number up to the last
 supported legacy version number.
 .El
+.It Xo
+.Nm
+.Cm version
+.Xc
+Displays the software version of the
+.Nm
+userland utility and the zfs kernel module.
+.El
 .El
 .Sh EXIT STATUS
 The following exit values are returned:
```

You're going to need to do something similar for `zfs(8)`.

TerraTech replied:
done

TerraTech commented on cmd/zfs/zfs_main.c:
```diff
@@ -8059,6 +8065,18 @@ zfs_do_project(int argc, char **argv)
+	if (zfs_version_print() == -1)
+		exit(errno);
+
+	exit(0);
```
gotcha

TerraTech replied:
done

TerraTech commented:
@behlendorf force-pushed requested changes.

rlaager commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1942,67 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+
+	if (close(fd) == -1)
+		return (-1);
+
```
extra newline

TerraTech replied:
done

rlaager commented on man/man8/zpool.8:
```diff
@@ -2430,7 +2439,6 @@ See
 .Xr zpool-features 5
 for details on compatibility with systems that support feature flags, but do not
 support all features enabled on the pool.
-.Bl -tag -width Ds
```
This removal (and the .El below) are wrong, as it then unindents the -a and -V. Bring them both back and this will be formatted correctly.

Put differently, relative to what you have here, apply this diff:
```
--- zpool.8.orig	2019-04-14 13:46:43.408967141 -0500
+++ zpool.8	2019-04-14 13:46:24.900746044 -0500
@@ -2439,6 +2439,7 @@
 .Xr zpool-features 5
 for details on compatibility with systems that support feature flags, but do not
 support all features enabled on the pool.
+.Bl -tag -width Ds
 .It Fl a
 Enables all supported features on all pools.
 .It Fl V Ar version
@@ -2448,6 +2449,7 @@
 flag is specified, no features will be enabled on the pool.
 This option can only be used to increase the version number up to the last
 supported legacy version number.
+.El
 .It Xo
 .Nm
 .Cm version
```

TerraTech replied:
@rlaager Can you please check zfs.8 as well to make sure it is sane?

TerraTech replied:
zpool.8 done

rlaager replied:
They both look good.

Lady-Galadriel commented:
This pull request seems to satisfy the intent behind this bug #2501 from @behlendorf.
I'll leave it to others for closure.

rlaager commented:
@TerraTech Assuming you're updating this to fix the man page change, you should add a line like this to the end of your commit message:
`Closes #2501`

TerraTech commented:
@rlaager rebased to master and force pushed

rlaager commented:
I applied this to my test 0.8.0 package, built on Ubuntu 18.10, and tested. All of the options (`-V`, `--version`, and `version`) work for both `zpool` and `zfs`. I tested as root and non-root.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1940,3 +1942,66 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
+void
+zfs_version_userland(char *version, int len)
+{
+	(void) strlcpy(version, ZFS_META_GITREV, len);
```
@TerraTech I thought this was ready to go, but we're going to need to make one last change.  Using `ZFS_META_GITREV` here results in `zfs --version` printing `"unknown"` for the user land version when building from a zfs tarball (no git tree).  For example:

```sh
$ zfs version
unknown
zfs-kmod-0.8.0-rc3
```

That's clearly no good.  Can you change this back to the version and release which is what is used by the kmod.   This way, the user and kernel version strings will be reported consistently.  We can always investigate further extending the command to provide the `ZFS_META_GITREV` if we so choose.  Here's the local change I made to test this.

```diff
diff --git a/lib/libzfs/libzfs_util.c b/lib/libzfs/libzfs_util.c
index 7e097c1012..c0ca266c95 100644
--- a/lib/libzfs/libzfs_util.c
+++ b/lib/libzfs/libzfs_util.c
@@ -55,7 +55,6 @@
 #include <zfs_fletcher.h>
 #include <libzutil.h>
 #include <sys/zfs_sysfs.h>
-#include "zfs_gitrev.h"
 
 int
 libzfs_errno(libzfs_handle_t *hdl)
@@ -1949,7 +1948,8 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, 
 void
 zfs_version_userland(char *version, int len)
 {
-       (void) strlcpy(version, ZFS_META_GITREV, len);
+       (void) strlcpy(version, "zfs-" ZFS_META_VERSION "-"
+           ZFS_META_RELEASE, len);
 }
 
 /*

```

behlendorf commented:
@TerraTech hopefully you don't mind, when applying this change I included the change to `ZFS_META_ALIAS` from `ZFS_META_GITREV` mentioned above.

TerraTech commented:
@behlendorf force-pushed and tried to provide for the best of both worlds.

TerraTech commented:
@behlendorf Can you take a look at my updated patch?  It will emit the right version regardless if via tarball or git tree.  Thanks!

TerraTech commented:
I have created a new PR #8632 that includes the follow up work that missed this PR's merge window.
